examples/end-to-end-examples/support_chatbot/app_slack.py

```diff
@@ -0,0 +1,333 @@
+import os
+import time
+import random
+import string
+from io import BytesIO
+import hashlib
+import hmac
+
+from argparse import ArgumentParser, Namespace
+from langchain.embeddings import MosaicMLInstructorEmbeddings
+from langchain.llms import MosaicML
+from chatbot import ChatBot
+
+from slack_sdk import WebClient
+from slack_sdk.errors import SlackApiError
+from flask import Flask, request, jsonify
+from pyngrok import ngrok
+
+from oci_converser import OCIObjectStorageManager
+
+ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
+app = Flask(__name__)
+
+BOT_USER_ID = "U05MKGP6J84"
+processed_events = set()
+
+REDIRECT_URI = 'http://localhost:3000/login/callback'
+
+def parse_args() -> Namespace:
+    """Parse commandline arguments."""
+    parser = ArgumentParser(description='Run a chatbot!')
+    parser.add_argument('--endpoint_url', type=str, default='https://models.hosted-on.mosaicml.hosting/mpt-30b-chat/v1/predict', required=False, help='The endpoint of our MosaicML LLM Model')
+    parser.add_argument('--max_length', type=int, default=6000, required=False, help='The maximum size of context from LangChain')
+    parser.add_argument('--chunk_size', type=int, default=1200, required=False, help='The chunk size when splitting documents')
+    parser.add_argument('--chunk_overlap', type=int, default=800, required=False, help='The overlap between chunks when splitting documents')
+    parser.add_argument('--retrieval_k', type=int, default=5, required=False, help='The number of chunks to retrieve as context from vector store')
+    parser.add_argument('--model_k', type=int, default=10, required=False, help='The number of outputs model should output')
+    parser.add_argument('--repository_urls', type=str, default='https://github.com/mosaicml/composer,https://github.com/mosaicml/streaming,https://github.com/mosaicml/examples,https://github.com/mosaicml/diffusion,https://github.com/mosaicml/llm-foundry', required=False, help='The GitHub repository URLs to download')
+    parser.add_argument('--data_collecting', type=bool, default=False, help='Where successful threads will be stored')
+    parser.add_argument('--slack_token', type=str, help='Slack Token')
+    parser.add_argument('--slack_signing_secret', type=str, help='Slack Signing Secret')
+    parser.add_argument('--oci_data_storage', type=str, default='oci://mosaicml-internal-checkpoints/support-bot-demo/slack-data', help='Where successful threads will be stored')
+    parser.add_argument('--complex_chat', type=bool, default=False, help='Where successful threads will be stored')
+
+    parsed = parser.parse_args()
+    if parsed.repository_urls is not None:
+        parsed.repository_urls = ''.join(str(parsed.repository_urls).split()).split(',')
+    return parsed
+
+@app.route('/slack/events', methods=['POST'])
+def slack_events():
+    data = request.json
+    
+    # Immediately respond to Slack's challenge
+    if "challenge" in data:
```

Comment:
Its a security measure where upon startup slack will send verify the auth key (might not be exactly that but somehting of that nature) ~~ it will always happen on startup